File: backend/src/main/java/com/hostel/backend/controller/ComplaintController.java
package com.hostel.backend.controller;

import com.hostel.backend.dto.*;
import com.hostel.backend.enums.ComplaintCategory;
import com.hostel.backend.enums.ComplaintPriority;
import com.hostel.backend.enums.ComplaintStatus;
import com.hostel.backend.service.ComplaintService;
import jakarta.validation.Valid;
import org.springframework.data.domain.Page;
import org.springframework.format.annotation.DateTimeFormat;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.*;

import java.time.LocalDate;

@RestController
@RequestMapping("/api/complaints")
public class ComplaintController {

    private final ComplaintService complaintService;

    public ComplaintController(ComplaintService complaintService) {
        this.complaintService = complaintService;
    }

    @PostMapping
    public ResponseEntity<ComplaintResponse> createComplaint(
            @Valid @RequestBody ComplaintRequest request,
            Authentication authentication) {
        return ResponseEntity.ok(
                complaintService.createComplaint(authentication.getName(), request));
    }

    @PutMapping("/{id}/assign")
    public ResponseEntity<ComplaintResponse> assignWorker(
            @PathVariable Long id,
            @RequestBody AssignWorkerRequest request,
            Authentication authentication) {
        return ResponseEntity.ok(
                complaintService.assignWorker(id, request.getWorkerId(), authentication.getName()));
    }

    @PutMapping("/{id}/reassign")
    public ResponseEntity<ComplaintResponse> reassignWorker(
            @PathVariable Long id,
            @RequestBody AssignWorkerRequest request,
            Authentication authentication) {
        return ResponseEntity.ok(
                complaintService.reassignWorker(id, request.getWorkerId(), authentication.getName()));
    }

    @PutMapping("/{id}/reject")
    public ResponseEntity<ComplaintResponse> rejectComplaint(
            @PathVariable Long id,
            @RequestBody RejectComplaintRequest request,
            Authentication authentication) {
        return ResponseEntity.ok(
                complaintService.rejectComplaint(id, request.getReason(), authentication.getName()));
    }

    @PutMapping("/{id}/status")
    public ResponseEntity<ComplaintResponse> updateStatus(
            @PathVariable Long id,
            @RequestBody UpdateComplaintStatusRequest request,
            Authentication authentication) {
        return ResponseEntity.ok(
                complaintService.updateStatus(id, request, authentication.getName()));
    }

    @PutMapping("/{id}/close")
    public ResponseEntity<ComplaintResponse> closeComplaint(
            @PathVariable Long id,
            @RequestBody(required = false) CloseComplaintRequest request,
            Authentication authentication) {
        return ResponseEntity.ok(
                complaintService.closeComplaint(id, request, authentication.getName()));
    }

    @GetMapping
    public ResponseEntity<Page<ComplaintResponse>> getComplaints(
            Authentication authentication,
            @RequestParam(defaultValue = "0")  int page,
            @RequestParam(defaultValue = "10") int size,
            @RequestParam(required = false) ComplaintStatus status,
            @RequestParam(required = false) String blockName,
            @RequestParam(required = false) ComplaintCategory category,
            @RequestParam(required = false) ComplaintPriority priority,
            @RequestParam(required = false) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDate dateFrom,
            @RequestParam(required = false) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDate dateTo) {
        return ResponseEntity.ok(
                complaintService.getComplaintsByRole(
                        authentication.getName(), page, size,
                        status, blockName, category, priority, dateFrom, dateTo));
    }

    @GetMapping("/dashboard")
    public ResponseEntity<DashboardStatsResponse> getDashboard(Authentication authentication) {
        return ResponseEntity.ok(
                complaintService.getDashboardStats(authentication.getName()));
    }

    @GetMapping("/dashboard/student")
    public ResponseEntity<StudentDashboardStatsResponse> getStudentDashboard(
            Authentication authentication) {
        return ResponseEntity.ok(
                complaintService.getStudentDashboardStats(authentication.getName()));
    }

    @GetMapping("/dashboard/worker")
    public ResponseEntity<WorkerDashboardStatsResponse> getWorkerDashboard(
            Authentication authentication) {
        return ResponseEntity.ok(
                complaintService.getWorkerDashboardStats(authentication.getName()));
    }

    @GetMapping("/{id}")
    public ResponseEntity<ComplaintResponse> getComplaintById(
            @PathVariable Long id,
            Authentication authentication) {
        return ResponseEntity.ok(
                complaintService.getComplaintById(id, authentication.getName()));
    }
}

File: backend/src/controllers/complaint.controller.js
const asyncHandler = require('../utils/asyncHandler');
const complaintService = require('../services/complaint.service');

const createComplaint = asyncHandler(async (req, res) => {
  res.json(await complaintService.createComplaint(req.user.email, req.body));
});

const assignWorker = asyncHandler(async (req, res) => {
  res.json(await complaintService.assignWorker(req.params.id, req.body.workerId, req.user.email));
});

const reassignWorker = asyncHandler(async (req, res) => {
  res.json(await complaintService.reassignWorker(req.params.id, req.body.workerId, req.user.email));
});

const rejectComplaint = asyncHandler(async (req, res) => {
  res.json(await complaintService.rejectComplaint(req.params.id, req.body.reason, req.user.email));
});

const updateStatus = asyncHandler(async (req, res) => {
  res.json(await complaintService.updateStatus(req.params.id, req.body, req.user.email));
});

const closeComplaint = asyncHandler(async (req, res) => {
  res.json(await complaintService.closeComplaint(req.params.id, req.body, req.user.email));
});

const disputeComplaint = asyncHandler(async (req, res) => {
  res.json(await complaintService.disputeComplaint(req.params.id, req.body, req.user.email));
});

const getComplaints = asyncHandler(async (req, res) => {
  const page = parseInt(req.query.page ?? '0', 10);
  const size = parseInt(req.query.size ?? '10', 10);
  const { status, blockName, category, priority, dateFrom, dateTo } = req.query;
  res.json(
    await complaintService.getComplaintsByRole(req.user.email, page, size, {
      status,
      blockName,
      category,
      priority,
      dateFrom,
      dateTo,
    })
  );
});

const getDashboard = asyncHandler(async (req, res) => {
  res.json(await complaintService.getDashboardStats(req.user.email));
});

const getStudentDashboard = asyncHandler(async (req, res) => {
  res.json(await complaintService.getStudentDashboardStats(req.user.email));
});

const getWorkerDashboard = asyncHandler(async (req, res) => {
  res.json(await complaintService.getWorkerDashboardStats(req.user.email));
});

const getComplaintById = asyncHandler(async (req, res) => {
  res.json(await complaintService.getComplaintById(req.params.id, req.user.email));
});

module.exports = {
  createComplaint,
  assignWorker,
  reassignWorker,
  rejectComplaint,
  updateStatus,
  closeComplaint,
  disputeComplaint,
  getComplaints,
  getDashboard,
  getStudentDashboard,
  getWorkerDashboard,
  getComplaintById,
};
